agent/serf.go

```diff
@@ -25,15 +26,92 @@ import (
 type serfClient interface {
 	// Members lists members of the serf cluster.
 	Members() ([]serf.Member, error)
-	// Stream subcribes the caller to the serf event stream.
-	// Filter can be used to restrict the events.
-	// Returns an opaque handle to be used with Stop.
-	Stream(filter string, eventc chan<- map[string]interface{}) (serf.StreamHandle, error)
 	// Stop cancels the serf event delivery and removes the subscription.
 	Stop(serf.StreamHandle) error
 	// Close closes the client.
 	Close() error
 	// Join attempts to join an existing serf cluster identified by peers.
 	// Replay controls if previous user events are replayed once this node has joined the cluster.
+	// Returns the number of nodes joined
 	Join(peers []string, replay bool) (int, error)
+	// UpdateTags will modify the tags on a running serf agent
+	UpdateTags(tags map[string]string, delTags []string) error
+}
+
+func newRetryingClient(clientConfig serf.Config) (*retryingClient, error) {
+	client, err := reinit(clientConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &retryingClient{
+		client: client,
+		config: clientConfig,
+	}, nil
+}
+
+// Members lists members of the serf cluster.
+func (r *retryingClient) Members() ([]serf.Member, error) {
+	if err := r.reinit(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return r.client.Members()
+}
+
+// Stop cancels the serf event delivery and removes the subscription.
+func (r *retryingClient) Stop(handle serf.StreamHandle) error {
+	if err := r.reinit(); err != nil {
+		return trace.Wrap(err)
+	}
+	return r.client.Stop(handle)
+}
+
+// Join attempts to join an existing serf cluster identified by peers.
+// Replay controls if previous user events are replayed once this node has joined the cluster.
+// Returns the number of nodes joined
+func (r *retryingClient) Join(peers []string, replay bool) (int, error) {
+	if err := r.reinit(); err != nil {
+		return 0, trace.Wrap(err)
+	}
+	return r.client.Join(peers, replay)
+}
+
+// UpdateTags will modify the tags on a running serf agent
+func (r *retryingClient) UpdateTags(tags map[string]string, delTags []string) error {
+	if err := r.reinit(); err != nil {
+		return trace.Wrap(err)
+	}
+	return r.client.UpdateTags(tags, delTags)
+}
+
+// Close closes the client
+func (r *retryingClient) Close() error {
+	if r.client.IsClosed() {
+		return nil
+	}
+	return r.client.Close()
+}
+
+func (r *retryingClient) reinit() error {
+	if !r.client.IsClosed() {
```

Comment:
Looking at https://github.com/hashicorp/serf/blob/master/client/rpc_client.go It looks to me like IsClosed is only set when RPCClient.Close is called, unless I'm missing it I don't see another path.

Are you sure IsClosed is set when we're encountering this issue? To me it looks like we should probably be receiving some sort of IO error, and may need to try and reconnect if io errors are encountered.

Reply:
Client will be closed whenever listen fails.
https://github.com/hashicorp/serf/blob/v0.8.1/client/rpc_client.go#L147:
```go
func ClientFromConfig(c *Config) (*RPCClient, error) {
	// ...
	go client.listen()
	// ...
}
func (c *RPCClient) listen() {
	defer c.Close()
	// ...
}
```

Reply:
Ok thanks... just wanted to be sure as on initial inspection it didn't look like it did :)